### PR TITLE
Feature/5193 Use right SOAP version for wrapping the message

### PR DESCRIPTION
**Pull request by jkosternl**
Renamed the branch, to match the right issue number.. lost the previous PR.. but not the commits.

### CODE REVIEW

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java [hunk 1]
```diff
@@ -131,43 +143,74 @@ public void checkForSoapFault(Message responseBody, Throwable nested) throws Sen
...
 		}
-		if (session!=null && StringUtils.isNotEmpty(soapNamespaceSessionKey)) {
+
+		if (session != null && StringUtils.isNotEmpty(soapNamespaceSessionKey)) {
 			session.put(soapNamespaceSessionKey, SoapVersion.NONE.namespace);
 		}
 		return allowPlainXml ? message : new Message(""); // could replace "" with nullMessage(), but then tests fail.
 	}
 
+	private String extractMessageWithTransformers(TransformerPool transformerS11, TransformerPool transformerS12, Message message, PipeLineSession session, String soapNamespaceSessionKey) throws IOException, TransformerException, SAXException {
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:161]: Je hebt nu redelijk wat argumenten, maar extractBodySoap11 en extractBodySoap12 zijn class fields, dus die hoef je niet mee te geven :)

File: core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java [hunk 2]
```diff
@@ -34,41 +41,119 @@ public void testUnwrap() throws Exception {
...
 		String actual = prr.getResult().asString();
 
 		TestAssertions.assertEqualsIgnoreCRLF(expected, actual);
 	}
 
+	@Test
+	public void testShouldKeepPipeSoapConfigurationVersionWhileWrapping() throws Exception {
+		pipe.setDirection(Direction.UNWRAP);
+		pipe.registerForward(new PipeForward("pipe2", "READY"));
+		pipe.setStoreResultInSessionKey("originalMessage_unwrapped");
```
Comment (nielsm5) [inline core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java:62]: Dit doet als het goed is niets?

File: core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java [hunk 3]
```diff
@@ -34,41 +41,119 @@ public void testUnwrap() throws Exception {
...
+		wrapPipeSoap11.configure();
+		wrapPipeSoap11.start();
+
+		// Arrange 1
+		String inputSoap12 = "<soapenv:Envelope xmlns:soapenv=\"" + SOAPConstants.URI_NS_SOAP_1_2_ENVELOPE + "\"><soapenv:Body>" + DEFAULT_BODY_END;
+
+		PipeLineSession pipeLineSession = new PipeLineSession();
+		// Act
+		PipeRunResult prr = doPipe(pipe, inputSoap12, pipeLineSession);
+		PipeRunResult pipeRunResult = doPipe(wrapPipeSoap11, prr.getResult(), pipeLineSession);
```
Comment (nielsm5) [inline core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java:83]: Ik snap niet goed wat hier gebeurd. Je krijgt `soap 1.2`, en geeft `soap 1.1` terug?
Op zich prima, de 2e pipe zet immers de `setSoapVersion` op `1.1`.

Vervolgens doe je alles opnieuw maar dan stuur, en return je `soap 1.1`.
Hoe vertaalt dit zich naar `ShouldKeepPipeSoapConfigurationVersionWhileWrapping` Dat hij altijd `soap 1.1` moet gebruiken?

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java [hunk 4]
```diff
@@ -56,25 +57,32 @@
...
 	private static final String EXTRACT_BODY_XPATH = "/soapenv:Envelope/soapenv:Body/*";
 	private static final String EXTRACT_HEADER_XPATH = "/soapenv:Envelope/soapenv:Header/*";
 	private static final String EXTRACT_FAULTCOUNTER_XPATH = "count(/soapenv:Envelope/soapenv:Body/soapenv:Fault)";
-	private static final String EXTRACT_FAULTCODE_XPATH = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/faultcode";
-	private static final String EXTRACT_FAULTSTRING_XPATH = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/faultstring";
+	private static final String EXTRACT_FAULTCODE_XPATH_SOAP11 = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/faultcode";
+	private static final String EXTRACT_FAULTCODE_XPATH_SOAP12 = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/soapenv:Code";
+	private static final String EXTRACT_FAULTSTRING_XPATH_SOAP11 = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/faultstring";
+	private static final String EXTRACT_FAULTSTRING_XPATH_SOAP12 = "/soapenv:Envelope/soapenv:Body/soapenv:Fault/soapenv:Reason";
+	public static final String SESSION_MESSAGE_SOAP_VERSION = "messageSoapVersion";
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:85]: `SOAP_VERSION_SESSION_KEY` vind ik prima, overgenomen. 
De value laat ik liever in stand. Ik begreep van Tim namelijk dat men in de XML configuratie ook zelf allerlei dingen zomaar op de sessie kan zetten. Ik wil niet dat daar per ongeluk een clash in komt, dus dat iemand denkt ik zet ook effe een waarde op 'SoapVersion'. Denk dat deze key dan unieker is. Eensch?

File: core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java (hunk 2 above)
Comment (jkosternl) [reply core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java:62]: Regel 62 bleek idd weg te kunnen. Thanks.

File: core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java (hunk 3 above)
Comment (jkosternl) [reply core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java:83]: Ah ja, inderdaad is het logischer om het dan andersom te testen. Heb het nu aangepast, dus:
- Check A) SOAP 1.2 bericht er in -> unwrap -> wrap pipe=1.1 -> assert 1.1
- Check B) SOAP 1.1 bericht er in -> unwrap -> wrap pipe=1.2 -> assert 1.2

File: core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java [hunk 5]
```diff
@@ -323,7 +357,7 @@ public void testUnwrapConditional(boolean expectUnwrap) throws Exception {
 		if (expectUnwrap) {
 			TestAssertions.assertEqualsIgnoreCRLF(expected, actual);
 		} else {
-			assertEquals(input, actual);
+			Assertions.assertEquals(input, actual);
```
Comment (tnleeuw) [inline core/src/test/java/nl/nn/adapterframework/pipes/SoapWrapperPipeTest.java:360]: Can this use static import?

**Review by tnleeuw — CHANGES_REQUESTED** [review]
Please also add tests for processing non-SOAP XML messages, with flag `allowPlainXml` both `false` and `true`.

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java [hunk 6]
```diff
@@ -131,43 +143,74 @@ public void checkForSoapFault(Message responseBody, Throwable nested) throws Sen
...
+			extractedMessage = transformerS12.transform(message.asSource());
+		} else {
+			extractedMessage = transformerS11.transform(message.asSource());
+			if (StringUtils.isNotEmpty(extractedMessage)) {
+				soapVersion = SoapVersion.SOAP11;
+			} else {
+				extractedMessage = transformerS12.transform(message.asSource());
+				if (StringUtils.isNotEmpty(extractedMessage)) {
+					soapVersion = SoapVersion.SOAP12;
+				}
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:175]: Yes, I agree. Good suggestion. Will make it work like this as well, and add unit tests for NONE as well.

[the remark this answers is not in the document]
Comment (jkosternl) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:175]: I just figured out that your suggestion is not possible. Because we're using the method `extractMessageWithTransformers` also for optional parts, such as fault  sections, which don't have to be available anyways. Then we can't conclude it is not SOAP. 
When "NONE" is set on purpose, I'll skip parsing, as you suggested above in the previous comment.

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java [hunk 7]
```diff
@@ -131,43 +143,74 @@ public void checkForSoapFault(Message responseBody, Throwable nested) throws Sen
...
 		return allowPlainXml ? message : new Message(""); // could replace "" with nullMessage(), but then tests fail.
 	}
 
+	private String extractMessageWithTransformers(TransformerPool transformerS11, TransformerPool transformerS12, Message message, PipeLineSession session, String soapNamespaceSessionKey) throws IOException, TransformerException, SAXException {
+		// If SOAP version is already determined in the session, directly use the SOAP 1.2 transformer
+		SoapVersion soapVersion = getSoapVersionFromSession(session);
+		String extractedMessage;
+		if (soapVersion == SoapVersion.SOAP12) {
+			extractedMessage = transformerS12.transform(message.asSource());
+		} else {
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:167]: Ja inderdaad. Goed punt. Aangepast!

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkosternl) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java:175]: I've thought up another improvement for this.. I'll add it in the next commit.

**Review by tnleeuw — APPROVED** [review]

File: core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nielsm5) [reply core/src/main/java/nl/nn/adapterframework/soap/SoapWrapper.java]: Ja de waarde is zeker uniek(er) voor nu denk ik prima, je zou ook `SoapWrapper.SoapVersion` kunnen gebruiken oid?
